firefox_puppeteer/api/places.py

```diff
@@ -0,0 +1,156 @@
+        """, script_args=[url])
+
+    def is_bookmark_star_button_ready(self):
+        """Checks if the status of the star-button is not updating.
+
+        :returns: True, if the button is ready
+        """
+        return self.marionette.execute_script("""
+          let button = window.BookmarkingUI;
+
+          return button.status !== button.STATUS_UPDATING;
+        """)
+
+    def restore_default_bookmarks(self):
+        """Restores the default bookmarks for the current profile."""
+        try:
+            return self.marionette.execute_async_script("""
+              Cu.import("resource://gre/modules/BookmarkHTMLUtils.jsm");
+              Cu.import("resource://gre/modules/Services.jsm");
+
+              // Default bookmarks.html file is stored inside omni.jar,
+              // so get it via a resource URI
+              let defaultBookmarks = 'resource:///defaults/profile/bookmarks.html';
+
+              let observer = {
+                observe: function (aSubject, aTopic, aData) {
+                  Services.obs.removeObserver(observer, "bookmarks-restore-success");
+
+                  marionetteScriptFinished(true);
+                }
+              };
+
+              // Fire off an initially import
+              Services.obs.addObserver(observer, "bookmarks-restore-success", false);
+              BookmarkHTMLUtils.importFromURL(defaultBookmarks, true);
+            """, script_timeout=10000)
+        except TimeoutException:
+            # TODO: In case of a timeout clean-up the registered topic
+            pass
+
+    # Browser history related helpers #
+
+    def remove_all_history(self):
+        """Removes all history items."""
+        try:
+            self.marionette.execute_async_script("""
+                Cu.import("resource://gre/modules/Services.jsm");
+
+                let hs = Cc["@mozilla.org/browser/nav-history-service;1"]
+                         .getService(Ci.nsIBrowserHistory);
+
+                let observer = {
+                  observe: function (aSubject, aTopic, aData) {
+                    Services.obs.removeObserver(observer, 'places-expiration-finished');
+
+                    marionetteScriptFinished(true);
+                  }
+                };
+
+                // Remove the pages, then block until we're done or until timeout is reached
+                Services.obs.addObserver(observer, 'places-expiration-finished', false);
+
+                hs.removeAllPages();
+            """, script_timeout=10000)
+        except TimeoutException:
+            # TODO: In case of a timeout clean-up the registered topic
+            pass
+
+    def wait_for_visited(self, urls, callback):
+        """Waits until all passed-in urls have been visited."""
+        # Bug 1121691: Needs observer handling support with callback first
+        # Until then we have to wait about 4s to ensure the page has been indexed
+        callback()
+        from time import sleep
+        sleep(4)
```

Comment:
No, triggering the navigate() method has to happen first. Then we have to wait 4s until those urls have been visited. Good catch regarding the test. Will do it in a bit.